packages/babel-helper-create-class-features-plugin/src/decorators.ts

```diff
@@ -1983,11 +1985,47 @@ function transformClass(
   path.insertBefore(classAssignments.map(expr => t.expressionStatement(expr)));
 
   if (needsDeclaraionForClassBinding) {
-    path.insertBefore(
-      t.variableDeclaration("let", [
-        t.variableDeclarator(t.cloneNode(classIdLocal)),
-      ]),
-    );
+    const classBindingInfo = scopeParent.getBinding(classIdLocal.name);
+    if (!classBindingInfo.constantViolations.length) {
+      // optimization: reuse the inner class binding if the outer class binding is not mutated
+      path.insertBefore(
+        t.variableDeclaration("let", [
+          t.variableDeclarator(t.cloneNode(classIdLocal)),
+        ]),
+      );
+    } else {
+      const classOuterBindingDelegateLocal = scopeParent.generateUidIdentifier(
+        "t" + classIdLocal.name,
+      );
+      const classOuterBindingLocal = classIdLocal;
+      path.replaceWithMultiple([
+        t.variableDeclaration("let", [
+          t.variableDeclarator(t.cloneNode(classOuterBindingLocal)),
+          t.variableDeclarator(classOuterBindingDelegateLocal),
+        ]),
+        t.blockStatement([
```

Comment:
We use a block statement to simulate the inner scope so that it can inherit the `Yield` and `Await` production parameter.